src/main.rs

```diff
@@ -194,30 +209,213 @@ fn test_lossy_unicode() {
     assert_eq!("�", bad_s);
 }
 
-// def output_complete_command(cmdname, args, description, output_list):
-//     comps = ['complete -c', cmdname]
-//     comps.extend(args)
-//     if description:
-//         comps.append('--description')
-//         comps.append(description)
-//     output_list.append(lossy_unicode(' ').join([lossy_unicode(c) for c in comps]))
-
-fn output_complete_command(
-    cmdname: &str,
-    args: impl Iterator<Item = String>,
+const MAX_DESCRIPTION_WIDTH: usize = 78;
+const TRUNCATION_SUFFIX: char = '…';
+
+fn char_len(string: &str) -> usize {
+    string.chars().count()
+}
+
+fn fish_options(options: &str, existing_options: &mut HashSet<String>) -> Vec<String> {
+    let mut out = vec![];
+
+    for option in regex!(r###"[ ,"=\[\]]"###).split(options) {
+        let option = regex!(r###"\[.*\]"###).replace_all(option, "");
+        let option = regex!(
+            r###"(?x)
+                ^ [ \t\r\n\[\](){}.,:!]
+                | [ \t\r\n\[\](){}.,:!] $
+            "###
+        )
+        .replace_all(&option, "");
+
+        if option == "-" || option == "--" {
+            continue;
+        }
+
+        if regex!(r###"[{}()]"###).is_match(&option) {
+            continue;
+        }
+
+        let (fish_opt, num_dashes) = if option.starts_with("--") {
+            ("l", 2)
+        } else if option.starts_with("-") {
+            (if option.len() == 2 { "s" } else { "o" }, 1)
+        } else {
+            continue;
+        };
+
+        let option = format!(
+            "-{} {}",
+            fish_opt,
+            // Direct indexing of `option` won't panic due to how `num_dashes`
+            // is calculated.
+            fish_escape_single_quote(String::from(&option[num_dashes..]))
+        );
+
+        if existing_options.insert(option.clone()) {
+            out.push(option);
+        }
+    }
+
+    out
+}
+
+#[test]
+fn test_fish_options() {
+    let output: Vec<String> = vec!["-s 'f'".into(), "-l 'force'".into()];
```

Comment:
Tiny nit, it's usually helpful to casual readers to name the _true_ answer to something more descriptive like `expected_output` or similar.